psh/wget/wget.c

```diff
@@ -0,0 +1,500 @@
+{
+	char *hdr = NULL, *tmp;
+
+	while (hdr == NULL) {
+		tmp = memchr(common.readptr, '\n', common.buf + common.len - common.readptr);
+		if (tmp == NULL) {
+			/* Carry on reading */
+			if (wget_buf() != 0) {
+				return NULL;
+			}
+			continue;
+		}
+
+		*(tmp - 1) = '\0';
+		hdr = common.readptr;
+		common.readptr = tmp + 1;
+	}
+
+	return hdr;
+}
+
+
+static size_t wget_parsehdrs(void)
+{
+	char *hdr = NULL, *endptr, *num;
+	size_t contentLen = 0;
+
+	/* Read headers until no more data is available or empty line is found */
+	do {
+		hdr = wget_hdrnext();
+		if (hdr == NULL) {
+			return -1;
+		}
+
+		if (strncasecmp(hdr, "content-length: ", 15) == 0) {
+			num = strrchr(hdr, ' ');
+			if (num == NULL || *(num + 1) == '\0') {
+				break;
+			}
+
+			num = num + 1;
+			contentLen = strtoul(num, &endptr, 0);
+			if (*endptr != '\0') {
+				break;
+			}
+		}
+
+		if (common.debug) {
+			printf("%s\n", hdr);
+		}
+	} while (*hdr != '\0');
+
+	return contentLen;
+}
+
+
+static char *wget_response(void)
+{
+	char *status = NULL, *tmp;
+	int bytes;
+
+	common.len = 0;
+	common.readptr = common.buf;
+	/* Read first line */
+	while (status == NULL) {
+		bytes = read(common.socket, common.buf + common.len, sizeof(common.buf) - common.len);
+		if (bytes <= 0) {
+			return NULL;
+		}
+
+		if (common.len == 0 && common.debug) {
+			printf("\n-------Response-------\n");
+		}
+
+		if (common.debug) {
+			printf("%*s", bytes, common.buf);
+		}
+		common.len += bytes;
+
+		tmp = memchr(common.buf, '\n', common.len);
+		/* We have not received a whole line yet, carry on reading */
+		if (tmp == NULL) {
+			common.readptr = common.buf + bytes;
+			continue;
+		}
+
+		/* Omit \r\n */
+		common.readptr = tmp + 1;
+
+		*(tmp - 1) = '\0';
+		tmp = strstr(common.buf, "HTTP");
+		if (tmp == NULL) {
+			return NULL;
+		}
+
+		tmp = strchr(tmp, ' ');
+		if (tmp == NULL) {
+			return NULL;
+		}
+
+		status = tmp + 1;
+	}
+
+
+	return status;
+}
+
+
+static void wget_cleanup(void)
+{
+	/* Cursor enable */
+	printf("\033[?25h");
+
+	fclose(common.outfile);
+	close(common.socket);
+	free(common.host);
+}
+
+
+int psh_wget(int argc, char **argv)
+{
+	const char *url;
+	const char *output = NULL;
+	char *status, *endptr;
+	size_t len;
+	int statnum;
+	int c;
+	struct timespec ts;
+	time_t begin, end, delta;
+
+	while ((c = getopt(argc, argv, "O:d")) != -1) {
+		switch (c) {
+			case 'O':
+				output = optarg;
+				break;
+			case 'd':
+				common.debug = 1;
+				break;
+			default:
+			case 'h':
+				wget_help();
+				return 0;
+		}
+	}
+
+	if (argc - optind != 1) {
+		fprintf(stderr, "URL missing!\n");
+		return 2;
+	}
+
+	url = argv[optind];
+
+	if (wget_parseUrl(url) != 0) {
+		return 2;
+	}
+
+	/* By default save the file to the current directory with the name from url */
+	if (output == NULL) {
+		output = common.filename;
+	}
+
+	common.outfile = fopen(output, "w+");
+	if (common.outfile == NULL) {
+		fprintf(stderr, "Fail to open file %s\n", output);
```

Comment:
You are only writing to a file.
```suggestion
	common.outfile = fopen(output, "w");
	if (common.outfile == NULL) {
		fprintf(stderr, "Fail to create file %s\n", output);
```

Reply:
I agree with the first fix - it should be `w` instead of `w+`, but I'll stay with `open`.